testutil/testnode/node_init.go

```diff
@@ -0,0 +1,254 @@
+	"github.com/tendermint/tendermint/config"
+	tmos "github.com/tendermint/tendermint/libs/os"
+	"github.com/tendermint/tendermint/types"
+	tmtime "github.com/tendermint/tendermint/types/time"
+)
+
+func fundKeyringAccounts(cdc codec.Codec, accounts ...string) (keyring.Keyring, []banktypes.Balance, []authtypes.GenesisAccount) {
+	kb := keyring.NewInMemory(cdc)
+	genAccounts := make([]authtypes.GenesisAccount, len(accounts))
+	genBalances := make([]banktypes.Balance, len(accounts))
+
+	for i, acc := range accounts {
+		rec, _, err := kb.NewMnemonic(acc, keyring.English, "", "", hd.Secp256k1)
+		if err != nil {
+			panic(err)
+		}
+
+		addr, err := rec.GetAddress()
+		if err != nil {
+			panic(err)
+		}
+
+		balances := sdk.NewCoins(
+			sdk.NewCoin(app.BondDenom, sdk.NewInt(99999999999999999)),
+		)
+
+		genBalances[i] = banktypes.Balance{Address: addr.String(), Coins: balances.Sort()}
+		genAccounts[i] = authtypes.NewBaseAccount(addr, nil, 0, 0)
+	}
+	return kb, genBalances, genAccounts
+}
+
+func collectGenFiles(tmCfg *config.Config, encCfg encoding.Config, pubKey cryptotypes.PubKey, nodeID, chainID, rootDir string) error {
+	genTime := tmtime.Now()
+
+	gentxsDir := filepath.Join(rootDir, "gentxs")
+
+	initCfg := genutiltypes.NewInitConfig(chainID, gentxsDir, nodeID, pubKey)
+
+	genFile := tmCfg.GenesisFile()
+	genDoc, err := types.GenesisDocFromFile(genFile)
+	if err != nil {
+		return err
+	}
+
+	appState, err := genutil.GenAppStateFromConfig(
+		encCfg.Codec,
+		encCfg.TxConfig,
+		tmCfg,
+		initCfg,
+		*genDoc,
+		banktypes.GenesisBalancesIterator{},
+	)
+	if err != nil {
+		return err
+	}
+
+	genDoc = &types.GenesisDoc{
+		GenesisTime:     genTime,
+		ChainID:         chainID,
+		Validators:      nil,
+		AppState:        appState,
+		ConsensusParams: genDoc.ConsensusParams,
+	}
+
+	if err := genDoc.ValidateAndComplete(); err != nil {
+		return err
+	}
+
+	return genDoc.SaveAs(genFile)
+}
+
+func initGenFiles(
+	state map[string]json.RawMessage,
+	codec codec.Codec,
+	genAccounts []authtypes.GenesisAccount,
+	genBalances []banktypes.Balance,
+	file,
+	chainID string,
+) error {
+	// set the accounts in the genesis state
+	var authGenState authtypes.GenesisState
+	codec.MustUnmarshalJSON(state[authtypes.ModuleName], &authGenState)
+
+	accounts, err := authtypes.PackAccounts(genAccounts)
+	if err != nil {
+		return err
+	}
+
+	authGenState.Accounts = append(authGenState.Accounts, accounts...)
+	state[authtypes.ModuleName] = codec.MustMarshalJSON(&authGenState)
+
+	// set the balances in the genesis state
+	var bankGenState banktypes.GenesisState
+	codec.MustUnmarshalJSON(state[banktypes.ModuleName], &bankGenState)
+
+	bankGenState.Balances = append(bankGenState.Balances, genBalances...)
+	state[banktypes.ModuleName] = codec.MustMarshalJSON(&bankGenState)
+
+	appGenStateJSON, err := json.MarshalIndent(state, "", "  ")
+	if err != nil {
+		return err
+	}
+
+	cparams := types.DefaultConsensusParams()
+	cparams.Block.TimeIotaMs = 1
```

Comment:
[Non Blocking][Question]
Doesn't it make sense to take the consensus params as a parameter for more flexibility?
Especially the block time, I guess it makes sense to give the user the abilitiy to change it.
Then, we can for example check if the provided consensus params are null, we initialize using the default ones.

Reply:
good idea 903bc0b96e40d319aff87ebf044dcecf164a4586